scene/gui/control.cpp

```diff
@@ -2443,9 +2452,15 @@ Ref<Texture2D> Control::get_theme_icon(const StringName &p_name, const StringNam
 		}
 	}
 
+	if (data.theme_icon_cache.has(p_theme_type) && data.theme_icon_cache[p_theme_type].has(p_name)) {
+		return data.theme_icon_cache[p_theme_type][p_name];
+	}
+
 	List<StringName> theme_types;
 	_get_theme_type_dependencies(p_theme_type, &theme_types);
-	return get_theme_item_in_types<Ref<Texture2D>>(data.theme_owner, data.theme_owner_window, Theme::DATA_TYPE_ICON, p_name, theme_types);
+	Ref<Texture2D> icon = get_theme_item_in_types<Ref<Texture2D>>(data.theme_owner, data.theme_owner_window, Theme::DATA_TYPE_ICON, p_name, theme_types);
+	data.theme_icon_cache[p_theme_type][p_name] = icon;
```

Comment:
Is this some HashMap magic that you can assign `p_name` without creating `p_theme_type` map?
I'd expect something like
```C++
if (!data.theme_icon_cache.has(p_theme_type)) {
    data.theme_icon_cache[p_theme_type] = Theme::ThemeIconMap>()
}
data.theme_icon_cache[p_theme_type][p_name] = icon;
```

Reply:
> Is this some HashMap magic

Yes, there is an implicit constructor, IIRC.